### PR TITLE
replace FS by RootFS

### PR DESCRIPTION
**Pull request by aluzzardi**
- Add support for `@deprecated` GraphQL directive in the API
- Add support for handling `@deprecated` fields in the Go SDK codegen
- Renamed `fs` -> `rootfs` and `withFS` -> `withRootfs`
- Kept backward compat with deprecated `fs` / `withFS` fields

@helderco @slumbering @dolanor: The same `@deprecated` support needs to be added to Python and NodeJS as well. We're going to have a bunch like this.

### CODE REVIEW

**Comment by aluzzardi** [comment]
Nice -- CLI is failing because I forgot to update the code and the Go linter is catching that:

```
Error: #17 87.65 docs/current/sdk/go/snippets/multiplatform-support/build-images-cross-compilation/main.go:84:16: SA1019: client.
[214](https://github.com/dagger/dagger/actions/runs/3483199821/jobs/5826422113#step:4:215)
#17 87.65 	Container(dagger.ContainerOpts{Platform: platform}).
[215](https://github.com/dagger/dagger/actions/runs/3483199821/jobs/5826422113#step:4:216)
#17 87.65 	WithFS is deprecated: replaced by withRootfs (staticcheck)
[216](https://github.com/dagger/dagger/actions/runs/3483199821/jobs/5826422113#step:4:217)
#17 87.65 		binaryCtr := client.
[217](https://github.com/dagger/dagger/actions/runs/3483199821/jobs/5826422113#step:4:218)
#17 87.65
```

File: core/schema/container.graphqls
Comment (aluzzardi) [inline core/schema/container.graphqls:30]: Let the bike shed commence.

I wasn't sure whether this should be `rootfs`, `rootFs` or `rootFS`.

This becomes `Rootfs()` and `WithRootfs()` in Go which is not ideal.

Which one works best in GraphQL and respective languages, @helderco @slumbering?

(note: if we were to call this `rootFs`, it would be converted as `rootFS` by Go codegen)

**Comment by aluzzardi** [comment]
@sipsma I think I might have hit a cache issue, false positive (!!).


```console
$  ./hack/make sdk:go:test  
[all good, even though it fails on CI]
[edit a random file to bust the cache]
$  ./hack/make sdk:go:test  
#22 go build -o ./bin/dagger-engine-session -ldflags -s -w /app/cmd/engine-session
Error: input:1: Cannot query field "rootfs" on type "Container".
$  ./hack/make sdk:go:test  
[all good]
```

File: internal/mage/util/util.go
Comment (aluzzardi) [inline internal/mage/util/util.go:119]: FYI @sipsma

File: core/schema/container.graphqls
Comment (helderco) [reply core/schema/container.graphqls:30]: In Python both `rootfs` and `root_fs` feel natural. I'm partial to `rootfs` but I see what you mean in Go.

**Comment by helderco** [comment]
@aluzzardi It's easy enough to add the deprecation notice but I have an issue with the reference documentation. In Sphinx a [version is required](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated). I may use a simple warning instead.

File: core/schema/container.graphqls
Comment (aluzzardi) [reply core/schema/container.graphqls:40]: Nice! Will do

Comment (helderco) [reply core/schema/container.graphqls:40]: Just noticed an issue... Since the reason references a GraphQL name, when you codegen as is you don't get it through the proper naming conventions.

File: core/schema/container.graphqls
Comment (aluzzardi) [reply core/schema/container.graphqls:30]: Thanks @helderco!

@sipsma @shykes Any thoughts on this?

- `rootfs` (graphql) ->`Rootfs()` (Go) -> `rootfs` (Python) -> ?? (Node)
- `rootFS` (graphql) -> `RootFS()` (Go) -> `root_fs` (Python) -> ?? (Node)

File: core/schema/container.graphqls
Comment (aluzzardi) [reply core/schema/container.graphqls:40]: Yeah, I noticed that as well when doing this.

I was thinking having a somewhat "wrong" name is better than not suggesting an alternative at all. And instead of picking the name in a given SDK, I just picked the GraphQL field name. WDYT?

Comment (aluzzardi) [reply core/schema/container.graphqls]: Implemented the suggestion above in the last commit -- @helderco thoughts?

File: core/schema/container.graphqls
Comment (sipsma) [reply core/schema/container.graphqls:30]: > rootfs (graphql) ->Rootfs() (Go) -> rootfs (Python) -> ?? (Node)

That's my vote, I think it's pretty typical to write "rootfs"/"Rootfs" as a single word rather than emphasizing that "FS" is a acronym. Also just slightly easier to type.

Not a strong opinion though, either is fine with me.

File: core/schema/container.graphqls
Comment (helderco) [reply core/schema/container.graphqls]: I like that it enables me to format and turn it into a link, but I'm wondering about the playground. If you go to the GraphQL API directly the `@` will look alien there.

Maybe if we have a set of formatted strings: `Replaced by <field>.` and any others that we may need. So an SDK regex could match against `Replaced by withRootfs.` and extract the field from that pattern.

Comment (aluzzardi) [reply core/schema/container.graphqls]: > I like that it enables me to format and turn it into a link, but I'm wondering about the playground. If you go to the GraphQL API directly the @ will look alien there.

Not *too* alien I think? Maybe \`field\` would look more natural?

> Maybe if we have a set of formatted strings: Replaced by <field>. and any others that we may need. So an SDK regex could match against Replaced by withRootfs. and extract the field from that pattern.

Not sure I fully understand this. What would we write as the reason in GraphQL and how would we know which field it is?

Comment (aluzzardi) [reply core/schema/container.graphqls]: Changed this to

```
Replaced by `withRootFs`
```

Looks just fine in raw GraphQL

Comment (helderco) [reply core/schema/container.graphqls]: > Not sure I fully understand this. What would we write as the reason in GraphQL and how would we know which field it is?

In GraphQL it would be just `Replaced by withRootfs`. Then in the SDKs you would match by the whole phrase like `Replaced by (.*).` instead of just the subject.

Anyway, I'm good with the backticks, I'm going to update Python now quickly so you don't get blocked on it.

File: codegen/generator/nodejs/templates/src/method_comment.ts.tmpl
Comment (aluzzardi) [inline codegen/generator/nodejs/templates/src/method_comment.ts.tmpl:2]: @dolanor Please take a look! Is this the correct pattern?

File: sdk/nodejs/api/client.gen.ts
Comment (aluzzardi) [inline sdk/nodejs/api/client.gen.ts:524]: @slumbering Is this the correct way to do this in TS?

**Review by aluzzardi — COMMENTED** [review]
Added support for NodeJS

File: core/schema/container.graphqls
Comment (aluzzardi) [reply core/schema/container.graphqls]: Oh, I see!

Well, I assumed this could be more dynamic. For instance, we're about to deprecate `Host().Workdir()`. Depreciation reason could be:

```
Use `directory` with `path` set to . instead.
```

**Review by sipsma — APPROVED** [review]
Go parts of the changes and the naming of the api lgtm!

**Comment by aluzzardi** [comment]
About to merge this

@helderco Can you make the `field` changes in a follow up?
@dolanor @slumbering I implemented the NodeJS part to get this unblocked, could you review/make a follow up if needed?

/cc @kpenfound @jpadams @vikram-dagger FYI of the change

**Review by shykes — APPROVED** [review]

File: sdk/nodejs/api/client.gen.ts
Comment (dolanor) [reply sdk/nodejs/api/client.gen.ts:524]: After investigating, yes it seems to be that format.

File: core/schema/container.graphqls
Comment (helderco) [reply core/schema/container.graphqls]: To be able to cross-reference in python it's simpler to assume backticks only for other **methods**:

```
Use `directory` with path set to '.' instead.
```
In another object:
```
Use `Container.directory` instead.
```

File: core/schema/container.graphqls
Comment (dolanor) [reply core/schema/container.graphqls:30]: For typescript, based on: https://www.typescriptlang.org/tsconfig#watch-watchFile
I would say it would be either `rootfs` or `rootFs`.
I would prefer `rootfs`, but I'll let @slumbering have the last word :)

File: codegen/generator/nodejs/templates/src/method_comment.ts.tmpl
Comment (dolanor) [reply codegen/generator/nodejs/templates/src/method_comment.ts.tmpl:2]: It would be.

Question is, what if there is no description but it is deprecated? Then we would generate nothing.

I guess we would need something like that:
```suggestion
  {{- if or .Description .DeprecationReason -}}
```

**Review by dolanor — COMMENTED** [review]
small comments after the fact.
Maybe the one on the comment generation should be fixed or we might not generate deprecation notice in case there is no description in the first place

File: sdk/nodejs/api/client.gen.ts
Comment (slumbering) [reply sdk/nodejs/api/client.gen.ts:524]: yes it is @aluzzardi

File: core/schema/container.graphqls
Comment (slumbering) [reply core/schema/container.graphqls:30]: We are using camelCase mostly. So I would go for:
rootfs (graphql) ->Rootfs() (Go) -> rootfs (Python) -> rootFs (so rootFS) (Node)

**Pull request merged**
